File: src/main/java/wraith/fabricaeexnihilo/modules/farming/PlantableItem.java
package wraith.fabricaeexnihilo.modules.farming;

import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.util.ActionResult;
import wraith.fabricaeexnihilo.util.Lazy;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PlantableItem extends Item {
    private final Lazy<Block[]> plants;
    
    public PlantableItem(Lazy<Block[]> plants, FabricItemSettings settings) {
        super(settings);
        this.plants = plants;
    }
    
    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        var world = context.getWorld();
        var plantPos = context.getBlockPos().offset(context.getSide());
        var shuffledPlants = new ArrayList<>(List.of(plants.get()));
        Collections.shuffle(shuffledPlants);
        for (var plant : shuffledPlants) {
            var state = plant.getPlacementState(new ItemPlacementContext(context));
            if (state == null)
                return ActionResult.PASS;
    
            if (state.canPlaceAt(world, plantPos)) {
                world.setBlockState(plantPos, state);
                var player = context.getPlayer();
                if (player != null && !player.isCreative()) {
                    context.getStack().decrement(1);
                }
                return ActionResult.SUCCESS;
            }
        }
        return super.useOnBlock(context);
    }
    
}

File: src/main/kotlin/exnihilofabrico/content/farming/PlantableItem.kt
package exnihilofabrico.content.farming

import exnihilofabrico.content.base.BaseItem
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.item.ItemUsageContext
import net.minecraft.util.ActionResult

open class PlantableItem(val plants: List<BlockState>, settings: Settings): BaseItem(settings) {

    constructor(plant: BlockState, settings: Settings): this(listOf(plant), settings)
    constructor(plant: Block, settings: Settings): this(listOf(plant.defaultState), settings)

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        if(context.world.isClient)
            return ActionResult.SUCCESS
        val plantPos = context.blockPos.offset(context.facing)
        for (plant in plants.shuffled()) {
            if(placementCheck(context) && plant.canPlaceAt(context.world, plantPos)){
                context.world.setBlockState(plantPos, plant)
                return ActionResult.SUCCESS
            }
        }
        return super.useOnBlock(context)
    }

    open fun placementCheck(context: ItemUsageContext): Boolean {
        return true
    }
}